cf-app/src/main/kotlin/space/crowdforce/controllers/ProjectController.kt

```diff
@@ -3,23 +3,215 @@ package space.crowdforce.controllers
+
+        activityService.updateActivity(activityId, userId, activity.name, activity.description, activity.endTime, activity.startTime)
+    }
+
+    @DeleteMapping("/{projectId}/activities/{activityId}", produces = [MediaType.APPLICATION_JSON_VALUE])
+    suspend fun deleteActivites(
+        @PathVariable("projectId") projectId: Int,
+        @PathVariable("activityId") activityId: Int,
+        principal: Principal?
+    ) {
+        if (principal == null)
+            throw RuntimeException("Unauthrized")
+
+        val userId = userService.getUserIdByName(principal.name) ?: throw RuntimeException("Unauthrized")
+
+        activityService.deleteActivity(activityId, userId)
+    }
+
+    @PostMapping("/{projectId}/activities", produces = [MediaType.APPLICATION_JSON_VALUE])
+    suspend fun addActivites(
+        @PathVariable("projectId") projectId: Int,
+        principal: Principal?,
+        @RequestBody activity: ActivityFormUI
+    ) {
+        if (principal == null)
+            throw RuntimeException("Unauthrized")
+
+        val userId = userService.getUserIdByName(principal.name) ?: throw RuntimeException("Unauthrized")
+
+        activityService.createActivity(userId, projectId, activity.name, activity.description, activity.startTime, activity.endTime)
+    }
+
+    @GetMapping("/{projectId}/activities/{activityId}/participants", produces = [MediaType.APPLICATION_JSON_VALUE])
+    suspend fun getParticipants(
+        @PathVariable("projectId") projectId: Int,
+        @PathVariable("activityId") activityId: Int
+    ): List<SubscriberUI> {
+        return activityService.getParticipants(activityId).map { map(it) }
+    }
+
+
+    @PutMapping("/{projectId}/activities/{activityId}/participants", produces = [MediaType.APPLICATION_JSON_VALUE])
+    suspend fun takePartActivity(
+        @PathVariable("projectId") projectId: Int,
+        @PathVariable("activityId") activityId: Int,
+        principal: Principal?
+    ) {
+        if (principal == null)
+            throw RuntimeException("Unauthrized")
+
+        val userId = userService.getUserIdByName(principal.name) ?: throw RuntimeException("Unauthrized")
+
+        activityService.takePart(userId, activityId)
+    }
+
+    @DeleteMapping("/{projectId}/activities/{activityId}/participants", produces = [MediaType.APPLICATION_JSON_VALUE])
+    suspend fun leftActivity(
+        @PathVariable("projectId") projectId: Int,
+        @PathVariable("activityId") activityId: Int,
+        principal: Principal?
+    ) {
+        if (principal == null)
+            throw RuntimeException("Unauthrized")
+
+        val userId = userService.getUserIdByName(principal.name) ?: throw RuntimeException("Unauthrized")
+
+        activityService.deleteParticipant(userId, activityId)
+    }
+
+    fun map(user: User): SubscriberUI =
+        SubscriberUI(user.name)
+
+    fun map(activity: Activity): ActivityUI {
```

Comment:
I think it's better to move these mapper methods to some converters

Reply:
I agree with you.